File: 0010-regular-expression-matching/0010-regular-expression-matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        # Top-Down Memoization
        cache = {}

        def dfs(i, j):
            # If the result for this state is already computed, return it
            if (i, j) in cache:
                return cache[(i, j)]

            # If both strings are fully matched
            if i >= len(s) and j >= len(p):
                return True

            # If the pattern is consumed but the string is not
            if j >= len(p):
                return False

            # Check if the current characters match
            match = i < len(s) and (s[i] == p[j] or p[j] == '.')

            # Handle '*' in the pattern
            if (j + 1) < len(p) and p[j + 1] == '*':
                # Two cases:
                # 1. We don't use '*' (skip two characters in the pattern)
                # 2. We use '*' (check the next character in the string if it matches)
                cache[(i, j)] = dfs(i, j + 2) or (match and dfs(i + 1, j))
                return cache[(i, j)]

            if match:
                # Proceed with the next characters in both the string and the pattern
                cache[(i, j)] = dfs(i + 1, j + 1)
                return cache[(i, j)]

            # If no match
            cache[(i, j)] = False
            return False

        # Start the recursion with both indices at 0
        return dfs(0, 0)

File: 0010-regular-expression-matching/0010-regular-expression-matching.java
import java.util.HashMap;
import java.util.Map;
class Solution 
{
    private Map<String, Boolean> cache;
    private String s;
    private String p;
    public boolean isMatch(String s, String p) 
    {
        this.cache = new HashMap<>();
        this.s = s;
        this.p = p;
        return dfs(0, 0);
    }

    private boolean dfs(int i, int j) 
    {
        String key = i + "," + j;
        if (cache.containsKey(key)) 
        {
            return cache.get(key);
        }
        if (i >= s.length() && j >= p.length()) 
        {
            return true;
        }
        if (j >= p.length()) {
            return false;
        }

        boolean match = i < s.length() && (s.charAt(i) == p.charAt(j) || p.charAt(j) == '.');
        if ((j + 1) < p.length() && p.charAt(j + 1) == '*') 
        {
            boolean result = dfs(i, j + 2) || (match && dfs(i + 1, j));
            cache.put(key, result);
            return result;
        }
        if (match) 
        {
            boolean result = dfs(i + 1, j + 1);
            cache.put(key, result);
            return result;
        }

        cache.put(key, false);
        return false;
    }
}